beacon-chain/cache/payload_id.go

```diff
@@ -1,94 +1,63 @@
-type ProposerPayloadIDsCache struct {
-	slotToProposerAndPayloadIDs map[[keyLength]byte][vpIdsLength]byte
-	sync.RWMutex
+// PayloadIDCache is a cache that keeps track of the prepared payload ID for the
+// given slot and with the given head root.
+type PayloadIDCache struct {
+	slotToPayloadID map[primitives.Slot]RootToPayloadIDMap
+	sync.Mutex
 }
 
-// NewProposerPayloadIDsCache creates a new proposer payload IDs cache.
-func NewProposerPayloadIDsCache() *ProposerPayloadIDsCache {
-	return &ProposerPayloadIDsCache{
-		slotToProposerAndPayloadIDs: make(map[[keyLength]byte][vpIdsLength]byte),
-	}
+// NewPayloadIDCache returns a new payload ID cache
+func NewPayloadIDCache() *PayloadIDCache {
+	return &PayloadIDCache{slotToPayloadID: make(map[primitives.Slot]RootToPayloadIDMap)}
 }
 
-// GetProposerPayloadIDs returns the proposer and payload IDs for the given slot and head root to build the block.
-func (f *ProposerPayloadIDsCache) GetProposerPayloadIDs(
-	slot primitives.Slot,
-	r [fieldparams.RootLength]byte,
-) (primitives.ValidatorIndex, [pIdLength]byte, bool) {
-	f.RLock()
-	defer f.RUnlock()
-	ids, ok := f.slotToProposerAndPayloadIDs[idKey(slot, r)]
+// PayloadID returns the payload ID for the given slot and parent block root
+func (p *PayloadIDCache) PayloadID(slot primitives.Slot, root [32]byte) (primitives.PayloadID, bool) {
+	p.Lock()
+	defer p.Unlock()
+	inner, ok := p.slotToPayloadID[slot]
 	if !ok {
-		return 0, [pIdLength]byte{}, false
+		return primitives.PayloadID{}, false
 	}
-	vId := ids[:vIdLength]
-
-	b := ids[vIdLength:]
-	var pId [pIdLength]byte
-	copy(pId[:], b)
-
-	return primitives.ValidatorIndex(bytesutil.BytesToUint64BigEndian(vId)), pId, true
+	pid, ok := inner[root]
+	if !ok {
+		return primitives.PayloadID{}, false
+	}
+	return pid, true
 }
 
-// SetProposerAndPayloadIDs sets the proposer and payload IDs for the given slot and head root to build block.
-func (f *ProposerPayloadIDsCache) SetProposerAndPayloadIDs(
-	slot primitives.Slot,
-	vId primitives.ValidatorIndex,
-	pId [pIdLength]byte,
-	r [fieldparams.RootLength]byte,
-) {
-	f.Lock()
-	defer f.Unlock()
-	var vIdBytes [vIdLength]byte
-	copy(vIdBytes[:], bytesutil.Uint64ToBytesBigEndian(uint64(vId)))
-
-	var bs [vpIdsLength]byte
-	copy(bs[:], append(vIdBytes[:], pId[:]...))
-
-	k := idKey(slot, r)
-	ids, ok := f.slotToProposerAndPayloadIDs[k]
-	// Ok to overwrite if the slot is already set but the cached payload ID is not set.
-	// This combats the re-org case where payload assignment could change at the start of the epoch.
-	var byte8 [vIdLength]byte
-	if !ok || (ok && bytes.Equal(ids[vIdLength:], byte8[:])) {
-		f.slotToProposerAndPayloadIDs[k] = bs
+// SetPayloadID updates the payload ID for the given slot and headroot
+// it also prunes older entries in the cache
+func (p *PayloadIDCache) Set(slot primitives.Slot, root [32]byte, pid primitives.PayloadID) {
+	p.Lock()
+	defer p.Unlock()
+	if slot > 1 {
+		p.prune(slot - 2)
```

Comment:
What is the rationale for pruning everything before `slot - 2` here? Just to double check, if the current slot is 32, entries for slots 0..29 will be pruned. Is this the correct behavior?

Reply:
it could be slot-1 perhaps, it's just to not keep an evergrowing cache. In the happy /normal case this will keep a map with only 1 entry.